proposed/swift-interop.md

```diff
@@ -0,0 +1,132 @@
+
+We plan to interop with Swift's Library Evolution mode, which brings an additional wrinkle into the Swift struct story. Swift's Library Evolution mode abstracts away all type layout and semantic information unless a type is explicitly marked as `@frozen`. In the Library Evolution case, all structs have "opaque" layout, meaning that their exact layout and category cannot be determined until runtime. As a result, we need to treat all "opaque" layout structs as possibly non-bitwise-movable at compile time as we will not know until runtime what the exact layout is. Swift/C++ interop does not use the Library Evolution mode, so it is not limited by opaque struct layouts. On the bright side, all "opaque" layout structs are passed as pointers in the calling convention, which helps simplify signature construction in the higher-level case. The size and layout information of a struct is available in its [Value Witness Table](https://github.com/apple/swift/blob/main/docs/ABIStabilityManifesto.md#value-witness-table), so we can look up this information at runtime for allocating struct instances and manipulating struct memory correctly.
+
+### Projecting Swift into .NET
+
+The majority of the work for Swift/.NET interop is determining how a type that exists in Swift should exist in .NET and what shape it should have. This section is a work in progress and will discuss how each feature in Swift will be projected into .NET, particularly in cases where there is not a corresponding .NET or C# language feature. Each feature should have a subheading describing how the projection will look and how any mechanisms to make it work will be designed.
+
+All designs in this section should be designed such that they are trimming and AOT-compatible by construction. We should work to ensure that no feature requires whole-program analysis (such as custom steps in the IL Linker) to be trim or AOT compatible.
+
+#### Swift to .NET Language Feature Projections
+
+TBD
+
+#### Projection Tooling Components
+
+The projection tooling should be split into these components:
+
+##### Importing Swift into .NET
+
+1. A tool that takes in a `.swiftinterface` file or Swift sources and produces C# code.
+2. A library that provides the basic support for Swift interop that the generated code builds on.
+3. User tooling to easily generate Swift projections for a given set of `.framework`s.
+4. (optional) A .NET shared framework package when targeting macOS, Mac Catalyst, iOS, or tvOS platforms that exposes the platform APIs or NuGet packages for each `.framework` that is exposed from Swift to .NET.
+    - This would be required to provide a single source of truth for Swift types so they can be exposed across an assembly boundary.
```

Comment:
You do not need shared framework for single source of truth. A regular package is enough.

We have not done shared framework for WinRT and Objective-C for good reasons (versioning and size), so I do not see why we would want to do it for Swift.

Reply:
I was intending for this to refer to a "FrameworkReference" that is automatically referenced as part of a TPM in a TFM, as we have done with WinRT and Objective-C projections on their relevant platforms. I'll re-word this to better explain my intent.